bus-mapping/src/circuit_input_builder.rs

```diff
@@ -287,20 +287,51 @@ impl<P: JsonRpcClient> BuilderClient<P> {
         Ok(Self {
             cli: client,
             chain_id: chain_id.into(),
-            // TODO: Get history hashes
-            history_hashes: Vec::new(),
         })
     }
 
-    /// Step 1. Query geth for Block, Txs and TxExecTraces
+    /// Step 1. Query geth for Block, Txs, TxExecTraces, history block hashes
+    /// and previous state root.
     pub async fn get_block(
         &self,
         block_num: u64,
-    ) -> Result<(EthBlock, Vec<eth_types::GethExecTrace>), Error> {
+    ) -> Result<(EthBlock, Vec<eth_types::GethExecTrace>, Vec<Word>, Word), Error> {
         let eth_block = self.cli.get_block_by_number(block_num.into()).await?;
         let geth_traces = self.cli.trace_block_by_number(block_num.into()).await?;
 
-        Ok((eth_block, geth_traces))
+        // fetch up to 256 blocks
+        let mut n_blocks = std::cmp::min(256, block_num as usize);
+        let mut next_hash = eth_block.parent_hash;
+        let mut prev_state_root: Option<Word> = None;
+        let mut history_hashes = vec![Word::default(); n_blocks];
+        while n_blocks > 0 {
+            n_blocks -= 1;
+
+            // TODO: consider replacing it with `eth_getHeaderByHash`, it's faster
+            let header = self.cli.get_block_by_hash(next_hash).await?;
+
+            // set the previous state root
+            if prev_state_root.is_none() {
+                prev_state_root = Some(header.state_root.to_word());
+            }
+
+            // latest block hash is the last item
+            let block_hash = header
+                .hash
+                .ok_or(Error::EthTypeError(eth_types::Error::IncompleteBlock))?
+                .to_word();
+            history_hashes[n_blocks] = block_hash;
```

Comment:
@davidnevadoc We may want to review it before merge once we clarify the usage of `history_hashes`. 👍